### ISSUE TITLE
Memory issues of DMRG with mpirun

### ISSUE THREAD

**Issue by 1iquidmoon** [block-hczhai/block2-preview#10]
Hi All,
I met with a problem when I was doing a DMRG-SCF run with block2.

```
ERROR: mpirun -n 40 /home/cuiys/.conda/envs/test/bin/block2main dmrg.conf > dmrg.out 2>&1
Traceback (most recent call last):
  File "/home/SCRATCH/PYSCF_59/dmrg_restart.py", line 41, in <module>
    mc.kernel(mo1)
  File "/home/cuiys/.conda/envs/test/lib/python3.9/site-packages/pyscf/mcscf/mc1step.py", line 812, in kernel
    _kern(self, mo_coeff,
  File "/home/cuiys/.conda/envs/test/lib/python3.9/site-packages/pyscf/mcscf/mc1step.py", line 349, in kernel
    e_tot, e_cas, fcivec = casscf.casci(mo, ci0, eris, log, locals())
  File "/home/cuiys/.conda/envs/test/lib/python3.9/site-packages/pyscf/mcscf/mc1step.py", line 834, in casci
    e_tot, e_cas, fcivec = casci.kernel(fcasci, mo_coeff, ci0, log)
  File "/home/cuiys/.conda/envs/test/lib/python3.9/site-packages/pyscf/mcscf/casci.py", line 546, in kernel
    e_tot, fcivec = casci.fcisolver.kernel(h1eff, eri_cas, ncas, nelecas,
  File "/home/cuiys/.conda/envs/test/lib/python3.9/site-packages/pyscf/dmrgscf/dmrgci.py", line 730, in kernel
    executeBLOCK(self)
  File "/home/cuiys/.conda/envs/test/lib/python3.9/site-packages/pyscf/dmrgscf/dmrgci.py", line 948, in executeBLOCK
    raise err
  File "/home/cuiys/.conda/envs/test/lib/python3.9/site-packages/pyscf/dmrgscf/dmrgci.py", line 943, in executeBLOCK
    check_call(cmd, cwd=DMRGCI.runtimeDir, shell=True)
  File "/home/cuiys/.conda/envs/test/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'mpirun -n 40 /home/cuiys/.conda/envs/test/bin/block2main dmrg.co
[input_output.zip](https://github.com/block-hczhai/block2-preview/files/8525294/input_output.zip)
nf > dmrg.out 2>&1' returned non-zero exit status 137.
```

dmrg.conf, dmrg.out and FCIDUMP can be found attached. Could you tell me where the problem is? Thanks in advance.

**Comment by hczhai** [comment]
Hi, The detailed error message was not shown in the above description.
Could you go to the DMRG runetime directory (where you can find ``dmrg.sconf``), and manually execute the command ``mpirun -n 40 /home/cuiys/.conda/envs/test/bin/block2main dmrg.conf > dmrg.out``, and see what error you can get?
Secondly, how many CPU cores do you have in the node where you are running this example?

**Comment by hczhai** [comment]
I had a look at your attached ``dmrg.out``, and it looks like it is a memory issue. You may need to set a larger memory request in your **job submission script** (not in the input file).
Also, you are using 40 ``mpi`` processors. The code will consume less memory if you use hybrid openMP/MPI rather than full MPI.

**Comment by 1iquidmoon** [comment]
[input_output.zip](https://github.com/block-hczhai/block2-preview/files/8525455/input_output.zip)
Sorry I forgot to attach these files. By the way, there are 40 cores in the node

**Comment by hczhai** [comment]
If you are running in only one node, is there a reason for not using 40 openMP threads (you are using 1 openMP threads now)?

**Comment by 1iquidmoon** [comment]
Thanks for your advice. Maybe there is a memory issue. I will set a smaller ```mc.fcisolver.memory``` to see if can run. By the way it runs normally with ```stackblock``` in the same node, maybe two programs have different strategies on allocating memory.

And I could not understand your question well. The part of calling dmrg in my pyscf input file can be found as follows:
```
lib.num_threads (40)
dmrgscf.settings.BLOCKEXE = os.popen("which block2main").read().strip()
dmrgscf.settings.MPIPREFIX = 'mpirun -n 40'
dmrgscf.settings.BLOCKSCRATCHDIR = os.path.join(os.getcwd(), str(os.getpid()))

mc = dmrgscf.DMRGSCF(mf1,32,47).x2c()
mc.chkfile = 'dmrg.chk'
mo1 = lib.chkfile.load(mc.chkfile, 'mcscf/mo_coeff')
mc.conv_tol = 1e-5
mc.conv_tol_grad = 1e-2
mc.max_memory = 150000
mc.fcisolver.memory= 5 
mc.fcisolver.maxM = 3000
mc.fcisolver.tol = 1e-6
mc.kernel(mo1)
mc.analyze()
``` 
Hopefully you can get some clues.

**Comment by hczhai** [comment]
Thanks for the detailed response. Yes, it is normal that ``block2`` costs more memory than ``StackBlock`` with ``mpirun -n 40``.
In your script ``lib.num_threads (40)`` cannot be directly used to control the number of threads used in ``block2`` (you have to explicitly set ``mc.fcisolver.threads``). And it is not okay to use 40 threads and 40 mpi processors **simultaneously**. If you have 40 cores in your node, you should use (1) 40 mpi processor, with 1 thread per processor; or (2) 1 mpi processor, with 40 threads per processor; or (3) 4 mpi processors, with 10 threads per processor (for example).

Namely, ``number of threads * number of mpi processors = number of CPU cores``.

To do (1) you can set:

    dmrgscf.settings.MPIPREFIX = 'mpirun -n 40'
    ...
    mc.fcisolver.threads = 1

To do (2) you can set:

    dmrgscf.settings.MPIPREFIX = ''
    ...
    mc.fcisolver.threads = 40

To do (3) you can set:

    dmrgscf.settings.MPIPREFIX = 'mpirun --bind-to none -n 4'
    ...
    mc.fcisolver.threads = 10

And (2) or (3) should be faster than (1). (1) also consumes the largest amount of memory.
(1) (2) and (3) also work for ``StackBlock`` (if you compile it correctly).

I suggest that you may try (2) or (3) to see if it can solve the memory problem or be faster.

**Comment by 1iquidmoon** [comment]
Thanks for your explanation. I think my case is just (1). When working with ```stackblock``` it seems that 2 GB per core is sufficient with mpirun -n 40. I am currently running two job to compare the efficiency for (1) and (2). Could you tell me which case is supposed to be more memory-consuming?

**Comment by 1iquidmoon** [comment]
By the way, I have some problems on setting memory. In case (1), whether ```mc.fcisolver.memory ``` represents the memory per core? And in case (2), whether  ```mc.fcisolver.memory ``` represents the memory in the whole node?

**Comment by hczhai** [comment]
> I am currently running two job to compare the efficiency for (1) and (2). Could you tell me which case is supposed to be more memory-consuming?

In ``block2``, (1) should be more memory-consuming.

``mc.fcisolver.memory`` represents the memory per mpi processor. So

> In case (1), whether ``mc.fcisolver.memory`` represents the memory per core?

Yes, because in case (1), one mpi processor has 1 core.

> And in case (2), whether ``mc.fcisolver.memory`` represents the memory in the whole node?

Yes, because in case (2), one mpi processor has 40 cores (the whole node).

Note 1: for (2), setting ``dmrgscf.settings.MPIPREFIX = 'mpirun -n 1'`` may lead to the wrong cost (but correct answer). You need to use ``dmrgscf.settings.MPIPREFIX = 'mpirun --bind-to none -n 1'`` or ``dmrgscf.settings.MPIPREFIX = ''``.

Note 2: The meaning of ``mc.fcisolver.memory`` is different in ``StackBlock`` and ``block2``. In ``StackBlock`` , ``mc.fcisolver.memory`` is roughly the upper limit of the total memory used in the code. And when it is not enough, you will see an error message. In ``block2``, only part of the memory is allocated in the stack and most of them are allocated dynamically ("dynamically" means that they are not controlled by the ``mc.fcisolver.memory`` parameter). So we have to differ two cases in ``block2``:

(a) When ``mc.fcisolver.memory`` is too small, you will see an explicit error message with texts ``exceeding allowed memory ...``. Then you should increase ``mc.fcisolver.memory`` to solve the problem.

(b) When the memory in the node is not enough for dynamical allocations, you will see some error message likes ``exited on signal 9 (Killed).`` or ``killed due to out of memory`` and it is killed by the job scheduler due to the high memory consumption (like what you see in your ``dmrg.out``). Then you need to increase the memory in the job submission script or use a node with a larger memory, or decrease ``mc.fcisolver.memory`` (see the next point for when decreasing this amount can help).

(b') For the case of ``StackBlock`` and ``block2``, if ``mc.fcisolver.memory`` is too large, it is normally not a problem because the code will actually not occupy that whole amount. It will only consume what it needs. The code can still run even when ``mc.fcisolver.memory`` is larger than the physical memory in the node. But if ``mc.fcisolver.memory`` is super large (like 4 * physical memory in the node), it may cause problems because even if the claimed memory is not actually used, the memory **address** space can be exhausted.

In summary, in ``block2``, the ``mc.fcisolver.memory`` should ideally be set smaller than ``mc.fcisolver.memory`` for ``StackBlock``. For example, you may set it to be half of the node memory. However, because of the (b') above, if you set it to be the total memory in the node, normally, it should still be okay.

**Comment by 1iquidmoon** [comment]
Many thanks for your reply. Now I understand some basic things on these settings, and I have some thoughts. Would you like to tell me whether these are true (In ```block2```):

In a specific calculation with a given mode (parallel and/or serial)
(1) There is a lower limit of memory requirement. If the setting value is smaller than that, an error will be caused.
(2) There is a roughly upper limit of memory requirement. If the setting value is (not much) larger than that, it can still run normally but would not be faster.
(3) The setting value between cases in (1) and (2) is available. In this range larger memory leads to better efficient.

Thanks again for your patients.

**Comment by hczhai** [comment]
Detailed analysis of memory/efficiency stuff is indeed complicated in ``block2``.

(1) looks okay, but (2) and (3) are not correct. In ``block2``, if the memory requirement cannot be satisfied, it will abort at some point. So it is either (a) memory is okay, it can run normally; or (b) memory is not okay, it will abort [but there are two limits, one is for stack memory, and another is for dynamic memory. The abort message of the two can be different]. There is no intermediate state (c) "run in a less efficient mode" (if you use the default settings).

The code will not automatically do (c). But there is indeed a "manual" way to do (c). There are some special keywords (can be added through ``mc.fcisolver.extraline``). If you use these keywords, the code can run in a less efficient mode but lower memory requirement. See keywords ``min_mpo_mem``, ``expt_algo_type``, ``qc_mpo_type `` and ``lowmem_noise`` in https://block2.readthedocs.io/en/latest/user/keywords.html. But for your case, up to now I do not think your memory problem is related to these keywords. It should be mainly due to the large number of mpi processors.

Let me also provide a concrete example for the memory model in ``block2``.

Assuming you run a DMRG with bond dimensions 500 and 1000 in the sweep schedule.

Assuming:
With bond dimension M = 500 and 1 mpi processor, the code requires 5 GB stack memory and 24 GB dynamic memory.
With bond dimension M = 500 and 4 mpi processor, the code requires 1.5 GB stack memory and 8 GB dynamic memory per mpi processor.
With bond dimension M = 1000 and 1 mpi processor, the code requires 20 GB stack memory and 96 GB dynamic memory.
With bond dimension M = 1000 and 4 mpi processor, the code requires 6 GB stack memory and 32 GB dynamic memory per mpi processor.

Then:
Case (1) 1 mpi processor, set ``mc.fcisolver.memory = 4 GB``, the node has 150 GB memory in total. Result: the code abort at M = 500 sweep with error message "exceeding allowed memory" (because 4 < 5).
Case (2) 1 mpi processor, set ``mc.fcisolver.memory = 10 GB``, the node has 150 GB memory in total. Result: the code abort at M = 1000 sweep with error message "exceeding allowed memory" (because 10 < 20).
Case (3) 1 mpi processor, set ``mc.fcisolver.memory = 22 GB``, the node has 28 GB memory in total. Result: the code abort at M = 500 sweep with error message "exited on signal 9 (Killed)" (because 28 < 5 + 24).
Case (4) 1 mpi processor, set ``mc.fcisolver.memory = 22 GB``, the node has 100 GB memory in total. Result: the code abort at M = 1000 sweep with error message "exited on signal 9 (Killed)" (because 100 < 20 + 96).
Case (5) 4 mpi processor, set ``mc.fcisolver.memory = 22 GB``, the node has 150 GB memory in total. Result: the code abort at M = 1000 sweep with error message "exited on signal 9 (Killed)" (because 150 < (6 + 32) * 4).
Case (6) 1 mpi processor, set ``mc.fcisolver.memory = 22 GB``, the node has 150 GB memory in total. Result: the code can run normally up to M = 1000.

In practice we do not know the exact requirement of dynamic memory. A rough estimate is 2 to 4 times the ``Hmem`` number printed in ``dmrg.out``.

We can discuss other ways to reduce memory cost when the openMP alone cannot solve the problem.

**Comment by 1iquidmoon** [comment]
Well I suppose my problem can be solved simply by running with adequate memory in the serial fashion. I will leave the issue open until a single DMRG-CI process is done. And I will keep you updated if I find anything abnormal.

```block2``` is a very cool code and I would like to make further exploration with it. Thanks again!

**Comment by hczhai** [comment]
Thanks for your quick feedback. The feedback and test cases are important for us to improve the quality of the code.

I looked at your ``dmrg.conf`` and tried to execute it in a 36-core node. I found the following:

(1) The high memory cost is totally due to the large number of mpi processors. With ``mc.fcisolver.threads = 36`` (serial mode), for M = 1000, ``block2`` only requires the node to have ``~ 15 GB`` memory in total for your FCIDUMP.

(2) When you use 40 mpi processors, the extra memory cost is due to MPO. You can see there is a line ``memory usage =  1.30 GB`` in your ``dmrg.out`` and this is mainly the memory cost of the MPO (per processor). In total these MPOs will cost ``40 * 1.3 GB = 52 GB`` in the node. Your total memory cost will be ``52 + DMRG cost``. But ``StackBlock`` does not use the concept of MPO, so in ``StackBlock`` there is no such 52 GB MPO cost.

(3) With 40 mpi processors, in ``block2`` it is also possible to remove this MPO ``52 GB`` cost using the keyword ``min_mpo_mem True`` (but this will make the calculation slightly more inefficient). I tried this keyword with 36 mpi processors and I got ``memory usage =  183 MB`` in ``dmrg.out``. So using this keyword the MPO cost is reduced from ``52 GB`` to ``0.183 * 40 = 7.32 GB``. Then probably you will be able to run the 40 mpi case in one node (though in practice the 1-mpi 40-thread should be much better due to efficiency, see the next point)

(4) Using 36 threads the code looks much faster than your case with 40 mpi processors.
(4.1) In your output for 40-mpi, I see the first 13 sweeps took 11858 sec.
(4.2) Using 36 threads and one mpi, it took 2972 sec (4x speed up).
(4.3) Using 36 threads and one mpi and the extra keyword ``singlet_embedding``, it took only 1062 sec (11x speed up).

So I suggest you add the following (only useful for spin-adapted DMRG with non-zero total spin):

    mc.fcisolver.extraline = [ "singlet_embedding" ]

See Page 2682 left column in *J. Chem. Theory Comput.* 2017, **13**, 2681−2695 and the references 24,31,57 in that paper for the technical details regarding singlet embedding.

Both ``StackBlock`` and ``CheMPS2`` **use** singlet embedding by default. And there are **no** keywords to switch that off. In ``StackBlock`` if you want to avoid singlet embedding you need to do some minor changes in the source code. In ``CheMPS2`` there is no way to turn off singlet embedding.

``block2`` **does not use** singlet embedding by default. If there is the keyword ``singlet_embedding`` in the input file, ``block2`` will use singlet embedding.

Singlet embedding may not always make things faster. But for your case it seems that it looks good.

**Comment by 1iquidmoon** [comment]
So I wonder that ```min_mpo_mem True``` becomes helpful when memory was extremely small. I have applied 'Singlet embedding' which provides observable acceleration in my run. Thanks for your kind suggestion!
